src/librustc_codegen_llvm/common.rs

```diff
@@ -260,9 +260,13 @@ impl ConstMethods<'tcx> for CodegenCx<'ll, 'tcx> {
                     Some(GlobalAlloc::Memory(alloc)) => {
                         let init = const_alloc_to_llvm(self, alloc);
                         if alloc.mutability == Mutability::Mut {
-                            self.static_addr_of_mut(init, alloc.align, None)
+                            self.static_addr_of_mut(
+                                init,
+                                alloc.align,
+                                Some(&ptr.alloc_id.to_string()),
```

Comment:
You should compute this only once for both arms, to avoid duplication.

Also, judging from the output, it should be `format!("alloc{}", ptr.alloc_id)` because it doesn't look like `to_string` adds that prefix (the test should say `@alloc1` not `@1`).

Oh and to limit this to situations where we see the IR at all, you need to only have the name be `Some` when `self.sess().opts.fewer_names` is `false` or something like that.

cc @oli-obk @bjorn3 Can we also make sure `static_addr_of(_mut)` with a `None` last argument doesn't set any name? Oh and the methods themselves aren't great, they should be something like `global(_mut)_from_value` and to avoid a LLVM-ism it should be changed to take a miri alloc and do `const_alloc_to_llvm` internally (with e.g. Cranelift you just encode the miri alloc directly, bytes+relocations are linker-friendly).

Reply:
I would prefer a `global(_mut)_from_alloc` over `global(_mut)_from_value`, as a lot of the constant handling can be shared.

Reply:
> Oh and to limit this to situations where we see the IR at all, you need to only have the name be Some when self.sess().opts.fewer_names is false or something like that.

I'm not seeing `self.sess().opts.fewer_names` but I think that we can test on `self.sess().fewer_names()`?

Reply:
```rust
pub fn fewer_names(&self) -> bool {
        let more_names = self.opts.output_types.contains_key(&OutputType::LlvmAssembly)
            || self.opts.output_types.contains_key(&OutputType::Bitcode);

        // Address sanitizer and memory sanitizer use alloca name when reporting an issue.
        let more_names = match self.opts.debugging_opts.sanitizer {
            Some(Sanitizer::Address) => true,
            Some(Sanitizer::Memory) => true,
            _ => more_names,
        };

        self.opts.debugging_opts.fewer_names || !more_names
    }
```

Reply:
or just use the `self.sess().opts.debugging_opts.fewer_names`?

Reply:
> Also, judging from the output, it should be format!("alloc{}", ptr.alloc_id) because it doesn't look like to_string adds that prefix (the test should say @alloc1 not @1).

I didn't realize we had a display impl different from the debug impl. `format!("{:?}", ptr.alloc_id)` will do the right thing. Maybe we should adjust the display impl?

Reply:
> I'm not seeing self.sess().opts.fewer_names but I think that we can test on self.sess().fewer_names()?

I too think `self.sess().fewer_names()` should be used.

Reply:
> Oh and to limit this to situations where we see the IR at all, you need to only have the name be Some when self.sess().opts.fewer_names is false or something like that.

So, I tried to condition the definition on a value of true for `fewer_names` using both `self.sess().opts.debugging_opts.fewer_name` and (in a separate build) `self.sess().fewer_names()`:

```rust
let formatted_id = match self.sess().opts.debugging_opts.fewer_names {
                          true => Some(format!("alloc{}", ptr.alloc_id)),
                          false => None,
                       };
```

and I am still seeing formatting without "alloc" with this condition:

```
@0 = private unnamed_addr constant <{ [4 x i8] }> zeroinitializer, align 2
```


when I remove the condition, I get this:

```
@alloc5.0 = private unnamed_addr constant <{ [4 x i8] }> zeroinitializer, align 2
```

Reply:
Where am I going wrong?

Reply:
You inverted the condition. It should be:

```rust
let formatted_id = if self.sess().fewer_names() {
    None
} else {
    Some(format!("alloc{}", ptr.alloc_id))
};
```

It should return `None` when `fewer_names()` is true, not when it is false.

Reply:
"remove this condition" is:

```rust

                        if alloc.mutability == Mutability::Mut {
                            self.static_addr_of_mut(init, alloc.align, Some(format!("alloc{}", ptr.alloc_id).as_str()))
                        } else {
                            self.static_addr_of(init, alloc.align, Some(format!("alloc{}", ptr.alloc_id).as_str()))
                        }
```

Reply:
> I didn't realize we had a display impl different from the debug impl. format!("{:?}", ptr.alloc_id) will do the right thing. Maybe we should adjust the display impl?

@oli-obk I will try this later today.  Will let you know

Reply:
> You inverted the condition.

Thanks!  I misunderstood what this indicates.

Reply:
Oliver was correct about the debug formatting.  This addresses the missing "alloc" substring.

I pushed all of these changes in 6789a825e0f70df3dc77d49ac8542557953bf262

Thank you all very much for your help here!